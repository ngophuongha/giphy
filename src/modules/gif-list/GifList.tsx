import InfiniteScroll from "react-infinite-scroll-component";
import LazyLoad from "react-lazyload";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import { GifItem } from "../../modules/gif-list/gif-item/GifItem";
import { GifListProps } from "./GifList.type";
import { Loader } from "../../components/loader";

export const GifList = ({
  gifList,
  onLoadMore,
  hasMoreItems = false,
}: GifListProps): JSX.Element => {
  return (
    <InfiniteScroll
      dataLength={gifList.data.length}
      next={onLoadMore}
      hasMore={hasMoreItems}
      loader={<Loader />}
      className="overflow-hidden"
      scrollThreshold={1}
      endMessage={
        <p style={{ textAlign: "center" }}>
          <b>Yay! You have seen it all</b>
        </p>
      }
    >
      <Row lg={4} md={3} sm={2} className="g-2" data-testid="gif-list">
        {gifList.data.map((card, index) => (
          <LazyLoad key={index} offset={[-50, 100]} placeholder={<Loader />}>
            <Col key={index} className="d-flex img-fluid">
              <GifItem
                id={card.id}
                imageSrc={card.images.fixed_width_downsampled.url}
                title={card.title}
                username={card.username}
                rating={card.rating}
              />
            </Col>
          </LazyLoad>
        ))}
      </Row>
    </InfiniteScroll>
  );
};
