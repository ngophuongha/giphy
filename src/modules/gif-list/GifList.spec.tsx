import { screen, render, waitFor } from "../../../test";

import { GifList } from "./GifList";
import { mockGifPage } from "./GifList.fixture";

describe("Gif List", () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetModules();
  });
  it("should match snapshot", async () => {
    const { container } = render(
      <GifList onLoadMore={jest.fn} gifList={mockGifPage} />
    );
    expect(container).toMatchSnapshot();
    await waitFor(() =>
      expect(
        document.getElementsByClassName("lazyload-wrapper")[0]
      ).toBeInTheDocument()
    );
  });

  it("should not show end message when having more items to show", async () => {
    const { container } = render(
      <GifList onLoadMore={jest.fn} gifList={[]} hasMoreItems={true} />
    );
    expect(container).toMatchSnapshot();
    await waitFor(() => expect(screen.queryByText(/all/i)).toBeNull());
  });
});
