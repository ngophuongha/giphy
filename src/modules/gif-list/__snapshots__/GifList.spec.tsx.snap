// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gif List should match snapshot 1`] = `
<div>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component overflow-hidden"
      style="height: auto; overflow: auto;"
    >
      <div
        class="g-2 row row-cols-xl-4 row-cols-md-3 row-cols-2"
        data-testid="gif-list"
      >
        <div>
          <div
            class="lazyload-wrapper "
          >
            <div
              class="mt-2 card"
            >
              <img
                class="card-img-top"
                src="placeholder-img.png"
              />
              <div
                class="card-body"
              >
                <div
                  class="placeholder-glow card-title h5"
                >
                  <span
                    class="col-6 placeholder"
                  />
                </div>
                <p
                  class="placeholder-glow card-text"
                >
                  <span
                    class="col-7 placeholder"
                  />
                   
                  <span
                    class="col-4 placeholder"
                  />
                   
                  <span
                    class="col-4 placeholder"
                  />
                   
                  <span
                    class="col-6 placeholder"
                  />
                   
                  <span
                    class="col-8 placeholder"
                  />
                </p>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="lazyload-wrapper "
          >
            <div
              class="mt-2 card"
            >
              <img
                class="card-img-top"
                src="placeholder-img.png"
              />
              <div
                class="card-body"
              >
                <div
                  class="placeholder-glow card-title h5"
                >
                  <span
                    class="col-6 placeholder"
                  />
                </div>
                <p
                  class="placeholder-glow card-text"
                >
                  <span
                    class="col-7 placeholder"
                  />
                   
                  <span
                    class="col-4 placeholder"
                  />
                   
                  <span
                    class="col-4 placeholder"
                  />
                   
                  <span
                    class="col-6 placeholder"
                  />
                   
                  <span
                    class="col-8 placeholder"
                  />
                </p>
              </div>
            </div>
          </div>
        </div>
        <div />
      </div>
      <p
        style="text-align: center;"
      >
        <b>
          Yay! You have seen it all
        </b>
      </p>
    </div>
  </div>
</div>
`;

exports[`Gif List should not show end message when having more items to show 1`] = `
<div>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component overflow-hidden"
      style="height: auto; overflow: auto;"
    >
      <div
        class="g-2 row row-cols-xl-4 row-cols-md-3 row-cols-2"
        data-testid="gif-list"
      >
        <div />
        <div />
        <div />
      </div>
      <div
        class="g-2 row row-cols-xl-4 row-cols-md-3 row-cols-2"
      >
        <div
          class="col"
        >
          <div
            class="mt-2 card"
          >
            <img
              class="card-img-top"
              src="placeholder-img.png"
            />
            <div
              class="card-body"
            >
              <div
                class="placeholder-glow card-title h5"
              >
                <span
                  class="col-6 placeholder"
                />
              </div>
              <p
                class="placeholder-glow card-text"
              >
                <span
                  class="col-7 placeholder"
                />
                 
                <span
                  class="col-4 placeholder"
                />
                 
                <span
                  class="col-4 placeholder"
                />
                 
                <span
                  class="col-6 placeholder"
                />
                 
                <span
                  class="col-8 placeholder"
                />
              </p>
            </div>
          </div>
        </div>
        <div
          class="col"
        >
          <div
            class="mt-2 card"
          >
            <img
              class="card-img-top"
              src="placeholder-img.png"
            />
            <div
              class="card-body"
            >
              <div
                class="placeholder-glow card-title h5"
              >
                <span
                  class="col-6 placeholder"
                />
              </div>
              <p
                class="placeholder-glow card-text"
              >
                <span
                  class="col-7 placeholder"
                />
                 
                <span
                  class="col-4 placeholder"
                />
                 
                <span
                  class="col-4 placeholder"
                />
                 
                <span
                  class="col-6 placeholder"
                />
                 
                <span
                  class="col-8 placeholder"
                />
              </p>
            </div>
          </div>
        </div>
        <div
          class="col"
        >
          <div
            class="mt-2 card"
          >
            <img
              class="card-img-top"
              src="placeholder-img.png"
            />
            <div
              class="card-body"
            >
              <div
                class="placeholder-glow card-title h5"
              >
                <span
                  class="col-6 placeholder"
                />
              </div>
              <p
                class="placeholder-glow card-text"
              >
                <span
                  class="col-7 placeholder"
                />
                 
                <span
                  class="col-4 placeholder"
                />
                 
                <span
                  class="col-4 placeholder"
                />
                 
                <span
                  class="col-6 placeholder"
                />
                 
                <span
                  class="col-8 placeholder"
                />
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
